core/src/main/java/com/linecorp/armeria/common/util/AsyncCloseableShutdownHooks.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
+
+import java.util.ArrayDeque;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ThreadFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A utility class for adding a task with closing a {@link AsyncCloseable} on shutdown.
+ */
+public final class AsyncCloseableShutdownHooks {
```

Comment:
- I realized this will work for any `AutoCloseable`s because we're now using `.close()` only. How about refactoring a little to accept `AutoCloseable` instead of `AsyncCloseable`?
- How about renaming to `ShutdownHooks` so we can expand this for registering other hooks as well in the future?
- Could you add `@UnstableApi`?